c10/cuda/CUDADeviceAssertion.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <c10/cuda/CUDAException.h>
+#include <c10/macros/Macros.h>
+
+namespace c10 {
+namespace cuda {
+
+#ifdef TORCH_USE_CUDA_DSA
+// Copy string from `src` to `dst`
+static __device__ void dstrcpy(char* const dst, const char* const src) {
+  int i = 0;
+  // Copy string from source to destination, ensuring that it
+  // isn't longer than `C10_CUDA_DEVICE_SIDE_MAX_STR_LEN-1`
+  for (; src[i] != '\0' && i < C10_CUDA_DEVICE_SIDE_MAX_STR_LEN - 1; i++) {
+    dst[i] = src[i];
+  }
+  // If the loop terminated before `i` meets the condition below, then
+  // `dst` ends with a null-terminator from `src`. Otherwise, the loop
+  // reached maximum iterations and we add our own null-terminator to
+  // `dst`
+  if (i == C10_CUDA_DEVICE_SIDE_MAX_STR_LEN - 1) {
+    *dst = '\0';
+  }
+}
+
+__device__ __noinline__ void dsa_add_new_assertion_failure(
+    DeviceAssertionsData* assertions_data,
+    const char* assertion_msg0,
+    const char* filename0,
+    const char* function_name0,
+    const int line_number0,
+    const uint32_t caller0,
+    const dim3 block_id,
+    const dim3 thread_id) {
+  // `assertions_data` may be nullptr if device-side assertion checking
+  // is disabled at run-time. If it is disabled at compile time this
+  // function will never be called
+  if (!assertions_data) {
+    return;
+  }
+
+  // Atomically increment so other threads can fail at the same time
+  const auto nid = atomicAdd(&(assertions_data->assertion_count), 1);
+
+  if (nid > C10_CUDA_DEVICE_SIDE_ASSERTION_COUNT) {
+    // At this point we're ran out of assertion buffer space
+    // we could print a message about this, but that'd get
+    // spammy if a lot of threads did it, so we just silently
+    // ignore any other assertion failures. In most cases the
+    // failures will all probably analogous anyway.
+
+    // Problematically, if we return right away, then thousands
+    // of threads will hit `__trap()` all at once before the
+    // 10s of threads writing errors to UVM finish doing so. The
+    // result is that we can't actually see error messages.
+    // Therefore, we delay until those threads are done writing.
+    while (assertions_data->assertion_failure_written <
+           assertions_data->assertion_count) {
+    }
```

Comment:
Wondering if we should have a timeout here 🤔

Reply:
So, as @andreaskoepf mentioned, in his original implementation, he doesn't bother trapping; instead there's a boolean field that we regularly check on kernel launch that is stand-in for "an assert has failed". The big benefit is that we don't trash the CUDA context on an assertion fail, which is nice for interactive usage. This would also have the side effect of solving the race condition that you're having the threads busy wait for.

Even if you don't actually implement this, it seems like you could just have the thread fall through and keep running the rest of the kernel. The only problem is to make sure you have a "safe" value for indexing, but as mentioned, this seems like we ought to do anyway.

Reply:
I'll give this a try.

Reply:
@ezyang : The problem is: how do you communicate to the host that something's gone wrong on the GPU?

Raising an assert on the GPU triggers a failure mode that gets caught by `C10_CHECK` on any API call or kernel launch. So the host recognizes there was an issue on the next interaction with the GPU.

Raising a flag requires the host to check UVM at each API call and kernel launch to achieve the same behaviour. If we model the GPU's write to UVM as expensive (due device-host communication) but occurring only when an assert fails and the host's read from UVM as being cheap, then this is probably fine.

Reply:
Andreas had some gadget for asynchronously communicating back to host, so the host tests were cheap and just delayed.

However, @ngimel made the case that delaying the errors is not exactly the best since the tensors are trashed. So maybe delaying as you did here is reasonable.

Reply:
Earlier I was hitting a problem where:
* `assert()` wasn't always causing the GPU to go into failed device mode reliably
* `__trap()` always worked but raises an exception in the host process
* The above seemed fragile

So making a mark in UVM memory to indicate failure and then having the threads quietly produce garbage seems fine as a way of working around the above and trying to preserve the host process.

I've looked at Andreas's code, but I haven't found the gadget. I'm assuming it'll be a write to memory by the device accompanied by checking said memory on the host.

Reply:
Hmmm, if the `assert` was meant to protect against, eg, an out-of-bounds array access, then continuing on through the kernel is the wrong action and an early exit is needed.

Reply:
@andreaskoepf, maybe @ngimel has more thoughts, but if there's a situation where we can perform out-of-bounds memory accesses and use `assert()` to stop execution before this, that's a situation where trash-and-check may provide inadequate safeguards.

I've observed that the hard `assert` seems to sometimes get skipped in ways I can't reliably reproduce, so I'm switching to the trash-and-check strategy anyway.

Reply:
I've now switched this to not using `assert()` at all, trying to early-exit assert-failed kernels where possible, and relying solely on UVM to communicate assertion failures.